RFC-0004-expression-eval.md

```diff
@@ -0,0 +1,80 @@
+
+A key feature of unfenced functions is that they can be used to implement constant folding.  This is because the function can be evaluated at query planning time, and the result can be used as a constant in the query plan.  This can be a significant performance improvement, as it can reduce the amount of work that needs to be done at query execution time.  For example, if an unfenced function is evaluated over a partition column, the result can be used to prune partitions before the query is executed.
+
+Currently in Presto C++, there is no constant folding support for functions which are written in C++.  In practice, constant folding works for functions which have a corresponding Java implementation, but not for functions which are implemented in C++.  It works by essentially just using a forked set of functions written in Java to evaluate the constant expressions, and then during execution relying on the C++ functions to evaluate the non-constant expressions.  This has a few drawbacks:
+
+1) It requires a forked set of functions to be written in Java, which is a maintenance burden.
+2) An unfenced function written in C++ must have a corresponding Java implementation in order to work properly.
+
+RFC-0003 proposed a mechanism to both add a function registry to allow C++ implemented functions to be exposed to the coordinator without a corresponding Java implementation, and to add constant folding support via remote function evaluation.  This RFC has a similar goal, but represents an improvement over that approach.
+
+In this RFC, we propose to optimize row expressions as a whole, not just function calls.  This has a few advantages:
+
+1) It allows for constant folding to be applied to expressions which are not function calls, such as special form expressions.
+2) It can reduce the overhead of constant folding by allowing the evaluation engine to evaluate the expression as a whole, rather than evaluating each function call individually.
+
+### [Optional] Goals
+
+* Allow for a pluggable implementation of expression evaluation and use this in the Presto optimizer for constant folding
+* Allow connector authors to use this plugin to optimize their own expressions (for example, for filter pushdown)
+
+### [Optional] Non-goals
+
+## Proposed Implementation
+
+A new SPI will be added to Presto to allow for a pluggable implementation of expression evaluation.  A default implementation of this will be added which delegates to the existing expresion evaluation code, but this may be overridden by a custom plugin.
+
+A new plugin will be implemented which will optimize row expressions by serializing them and sending them to the Presto sidecar for evaluation.  The Presto sidecar will have a new endpoint which can take in a serialized row expression and return the result of evaluating it.
+
+> Endpoint: /v1/evaluate
+> 
+> HTTP verb: POST
+> 
+>Body: JSON list of serialized row expressions
+>
+> Response: JSON list of optimized serialized row expressions
```

Comment:
I expected the response to be a list of constant values that represent the results of evaluating the expressions. Is this not the case? Why are you saying "optimized ... expressions" and not "values"?

Reply:
For most cases, they will be values, however I think optimized expression better conveys that there are cases when they won't be simple constants (as I somehow understand from the word value).  Some examples I have in mind:

* CAST(int_column AS INT) => int_column
  * We would expect the expression optimizer to remove the unneeded cast
* IF(condition, 123, 0/0) => CAST(FAIL(...) AS INTEGER) -- also, similar for CASE statement
  * We would expect the expression optimizer to only fail the query if the false condition is reached
* COALESCE(val1, val2, val1, val3) => COALESCE(val1, val2, val3)
  * We would expect the expression optimizer to remove the unneeded and redundant statements in the coalesce

I planned to add the mechanism of how we determine what may be constant folded, and how we batch them and send them to the sidecar, as a part of this document (I am returning from PTO and wanted to get this out in draft state before I left).

Reply:
@tdcmeehan Tim, thank you for clarifying. I feel this is going beyond constant folding and it not supported in Velox / Prestissimo. I wonder if it makes sense to reduce the scope of this proposal to just constant folding, e.g. taking an expression with no column references and returning a value.

Reply:
While we wanted to push down as much as possible to the underlying evaluation engine, since we don't make any distinction between general expression optimization and the more specific action of reducing constant expressions to literal values, I understand that this might be considered to be out of scope for Velox, and we can make this distinction.

Reply:
@mbasmanova Consider the following expressions with references to other columns:

* `ABS(COALESCE(123, column1))` => `123`
* `IF(true, 123, column2)` => `123`

Do we consider the simplification of these expressions within scope for constant folding in Velox?

Reply:
I'm wondering if it's possible to avoid having to make multiple passes for each expression in the case of nested expressions which are able to constant fold into values.

If we take the above examples and combine them in a way:

`CppFlipSign(IF(MyCppFunc(COALESCE(123, column1)) < 456, 456, column2))`

Then whether or not Velox supports constant folding of special form expressions where there is a reference to a different column seems to make a big difference.  If Velox is able to constant fold those examples, then we can send the whole expression in one call.  If it can't, then it would require multiple passes between Presto and Velox to constant fold the expression.

Presto: `COALESCE(123, column1)` => `123` (contains a column reference, yet because of the properties of this special form expression, we can infer that we can just collapse it to 123)
Velox: `MyCppFunc(123)` => `123` (suppose for simplicity this just returns the input)
Presto: `IF(123 < 456, 456, column2)` => `456` (contains a column reference, yet because of the properties of this special form expression, we can infer that we can just collapse it to 456)
Velox: `CppFlipSign(456)` => `-456`

In this case, at least two calls are unavoidable, because Velox cannot determine that the special form expressions are constant foldable, and Presto can't execute the function by design.  I briefly looked at the code in Velox, can confirmed that it appears we only check if something is constant foldable if all of the inputs are constant expressions.  However, I'm wondering if we should add more sophistication to this process?  I am guessing that the original motivation behind constant folding expressions with only constant inputs would also apply to inferring the unique properties of special form expressions and constant folding such expressions where possible?

Reply:
@tdcmeehan Tim, these are all good questions / points. My initial thinking is to restrict Velox to "true" constant folding, e.g. evaluating expressions that do not reference any columns and producing a "value" as a result. I see how this may result in extra trips between coordinator and side-car, but I feel the alternative adds a lot of complexity. If you feel it is important to push this logic into the side-car, than, perhaps, a next step would be to describe all the optimizations we would expect to implement and then consider whether it makes sense to do that in Velox (or Prestissimo layer) or not. 

What do you think?

Reply:
I think this design works best when there's only a single hop, so it seems like the question is do we do this in Prestissimo or Velox.  I'll mention the optimizations needed to achieve general constant folding support which is at parity with Java (which includes expressions that contain column references), and leave it as a detail to be determined where the code lives.

Reply:
Sounds good. We may need to build some optimizer logic in C++ in Prestissimo to make this work. Note that conversion of expressions to Velox is one-way. It is not possible to go back cleanly.

Reply:
I don't think multiple passes in this use case is really an issue performance-wise.  Making the constant-folding in Velox more powerful might be useful in other use cases on execution side though.